clients/src/main/java/org/apache/kafka/common/internals/KafkaFutureImpl.java

```diff
@@ -140,12 +140,21 @@ R await(long timeout, TimeUnit unit)
      * futures's result as the argument to the supplied function.
      */
     @Override
-    public <R> KafkaFuture<R> thenApply(Function<T, R> function) {
+    public <R> KafkaFuture<R> thenApply(FunctionInterface<T, R> function) {
         KafkaFutureImpl<R> future = new KafkaFutureImpl<R>();
         addWaiter(new Applicant<>(function, future));
         return future;
     }
 
+    /**
+     * @See KafkaFutureImpl#thenApply(FunctionInterface)
+     */
+    @Deprecated
+    @Override
+    public <R> KafkaFuture<R> thenApply(Function<T, R> function) {
+        return thenApply((FunctionInterface<T, R>) function);
+    }
+
     @Override
     protected synchronized void addWaiter(BiConsumer<? super T, ? super Throwable> action) {
```

Comment:
For this to be useful to me in my work on KIP-183 I also need `addWaiter()` to be made `public` on `KafkaFuture`.

Reply:
Ok, I will take it in this commit.